utils/remote-run

```diff
@@ -59,31 +59,111 @@ class CommandRunner(object):
+        if "banner line contains invalid characters" in stderr:
+            return True
+
+        if "Connection to localhost closed by remote host" in stderr:
+            return True
+
+        if "kex_exchange_identification: Connection closed by remote host" in stderr:
+            return True
+
+        if "rsync error: unexplained error" in stderr:
+            return True
+
+        # Fallthrough. The error is not known and shouldn't be retried
+        return False
+
     def run_remote(self, command, remote_env={}):
         env_strings = ['{0}={1}'.format(k,v) for k,v in sorted(remote_env.items())]
         remote_invocation = self.remote_invocation(
             ['/usr/bin/env'] + env_strings + command)
-        remote_proc = self.popen(remote_invocation, stdin=subprocess.PIPE,
-                                 stdout=None, stderr=None)
-        if self.dry_run:
-            return
-        _, _ = remote_proc.communicate()
-        if remote_proc.returncode:
-            # FIXME: We may still want to fetch the output files to see what
-            # went wrong.
+        attempt = 1
+        remote_proc = None
+
+        while attempt <= 3:
+            remote_proc = self.popen(
+                remote_invocation,
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            )
+
+            if self.dry_run:
+                return
+
+            stdout, stderr = remote_proc.communicate()
+            stdout = stdout.decode('utf-8')
+            stderr = stderr.decode('utf-8')
+
+            # Print stdout to screen
+            print(stdout, end='')
+
+            # This is a transient and random error, and if this occurs, we should simply retry our ssh command.
+            if self.should_remote_command_retry(stderr):
+                attempt += 1
+                continue
+
+            print(stderr, end='', file=sys.stderr)
+
+            # Process error code
+            if remote_proc.returncode:
+                # FIXME: We may still want to fetch the output files to see what
+                # went wrong.
+                sys.exit(remote_proc.returncode)
+            else:
+                # Nothing went wrong. Return
+                return
+
+        if attempt > 3 and remote_proc is not None:
             sys.exit(remote_proc.returncode)
 
     def run_rsync(self, invocation, sources):
-        rsync_proc = self.popen(invocation,
-                                stdin=subprocess.PIPE,
-                                stdout=None, stderr=None)
-        if self.dry_run:
-            return
-        sources = '\n'.join(sources)
-        if self.verbose:
-            print(sources, file=sys.stderr)
-        _, _ = rsync_proc.communicate(sources.encode('utf-8'))
-        if not self.ignore_rsync_failure and rsync_proc.returncode:
+        attempt = 1
+        rsync_proc = None
+
+        while attempt <= 3:
+            rsync_proc = self.popen(
```

Comment:
This feels like it might make sense to refactor and share so we have a retryable command.

Reply:
These particular bits aren't really my changes; we had some of these internally and I thought it best to make the public version match the internal version here.